### PR TITLE
fix: Middleware param rewrite

### PR DESCRIPTION
**Pull request by balazsorban44**
Adds a test case for #35227

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: f50d62cc2504a139872404c912e0d2e4dc03ad41

`yarn testheadless test/e2e/middleware-param-rewrite-navigation/index.test.ts` 
- Middleware param rewrite navigation > should work

<details>
<summary>Expand output</summary>

  ● Middleware param rewrite navigation › should work

    expect(received).toBe(expected) // Object.is equality

    Expected: "b"
    Received: ""

      22 |   it('should work', async () => {
      23 |     // First render
    > 24 |     expect(await browser.elementByCss('span').text()).toBe('b')
         |                                                       ^
      25 |
      26 |     // After client navigation
      27 |     await browser.elementByCss('a').click()

      at Object.<anonymous> (e2e/middleware-param-rewrite-navigation/index.test.ts:24:55)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| buildDuration | 17.7s | 18s | ⚠️ +249ms |
| buildDurationCached | 6.9s | 6.9s | ✓ |
| nodeModulesSize | 372 MB | 372 MB | ⚠️ +4 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.502 | 3.531 | ⚠️ +0.03 |
| / avg req/sec | 713.86 | 707.97 | ⚠️ -5.89 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.558 | 1.572 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1604.49 | 1590.71 | ⚠️ -13.78 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.09 kB | 5.09 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.8 kB | 14.8 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

```diff
@@ -4510,7 +4510,7 @@
           result: interpolatedRoute
         };
       }
-      function omitParmsFromQuery(query, params) {
+      function omitParamsFromQuery(query, params) {
         var filteredQuery = {};
         Object.keys(query).forEach(function(key) {
           if (!params.includes(key)) {
@@ -4584,7 +4584,7 @@
               interpolatedAs = (0, _formatUrl).formatWithValidation({
                 pathname: result,
                 hash: finalUrl.hash,
-                query: omitParmsFromQuery(query, params)
+                query: omitParamsFromQuery(query, params)
               });
             }
           }
@@ -5461,7 +5461,7 @@
                               as = (0, _formatUrl).formatWithValidation(
                                 Object.assign({}, parsedAs1, {
                                   pathname: interpolatedAs.result,
-                                  query: omitParmsFromQuery(
+                                  query: omitParamsFromQuery(
                                     query,
                                     interpolatedAs.params
                                   )
```
</details>
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-53b0b458e260ed10.js"
+      src="/_next/static/chunks/main-94fabb2ceb9da9c9.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-53b0b458e260ed10.js"
+      src="/_next/static/chunks/main-94fabb2ceb9da9c9.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-53b0b458e260ed10.js"
+      src="/_next/static/chunks/main-94fabb2ceb9da9c9.js"
       defer=""
     ></script>
     <script
```
</details>
</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| buildDuration | 21.7s | 21.6s | -182ms |
| buildDurationCached | 7s | 7s | -17ms |
| nodeModulesSize | 372 MB | 372 MB | ⚠️ +4 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.495 | 3.667 | ⚠️ +0.17 |
| / avg req/sec | 715.3 | 681.77 | ⚠️ -33.53 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.604 | 1.67 | ⚠️ +0.07 |
| /error-in-render avg req/sec | 1558.91 | 1496.91 | ⚠️ -62 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.23 kB | 5.23 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15 kB | 15 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/middleware-param-rewrite | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

```diff
@@ -4510,7 +4510,7 @@
           result: interpolatedRoute
         };
       }
-      function omitParmsFromQuery(query, params) {
+      function omitParamsFromQuery(query, params) {
         var filteredQuery = {};
         Object.keys(query).forEach(function(key) {
           if (!params.includes(key)) {
@@ -4584,7 +4584,7 @@
               interpolatedAs = (0, _formatUrl).formatWithValidation({
                 pathname: result,
                 hash: finalUrl.hash,
-                query: omitParmsFromQuery(query, params)
+                query: omitParamsFromQuery(query, params)
               });
             }
           }
@@ -5461,7 +5461,7 @@
                               as = (0, _formatUrl).formatWithValidation(
                                 Object.assign({}, parsedAs1, {
                                   pathname: interpolatedAs.result,
-                                  query: omitParmsFromQuery(
+                                  query: omitParamsFromQuery(
                                     query,
                                     interpolatedAs.params
                                   )
```
</details>
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-53b0b458e260ed10.js"
+      src="/_next/static/chunks/main-94fabb2ceb9da9c9.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-53b0b458e260ed10.js"
+      src="/_next/static/chunks/main-94fabb2ceb9da9c9.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-53b0b458e260ed10.js"
+      src="/_next/static/chunks/main-94fabb2ceb9da9c9.js"
       defer=""
     ></script>
     <script
```
</details>
</details>
Commit: f50d62cc2504a139872404c912e0d2e4dc03ad41

File: test/e2e/middleware-param-rewrite-navigation/pages/a.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { useRouter } from 'next/router'
+import Link from 'next/link'
+
+export default function IndexPage() {
+  const router = useRouter()
+  return (
+    <>
+      <span>{router.query.param}</span>
+      <Link href="/b">
+        <a>To "b"</a>
```
Comment (ijjk) [inline test/e2e/middleware-param-rewrite-navigation/pages/a.js:10]:
```suggestion
        <a>To "a"</a>
```

**Review by ijjk — COMMENTED** [review]
Looks like we might need to ensure the query updating is triggered correctly when the page is static and it matches a `_middleware` route here https://github.com/vercel/next.js/blob/2b13c2f7fa6682a9b68b3a86589f2cef9de8d0e0/packages/next/client/index.tsx#L106 and then we also need to ensure the preflight request is triggered correctly for this case here https://github.com/vercel/next.js/blob/f14f62838aa080d40bbff124f17d698e55b99e62/packages/next/shared/lib/router/router.ts#L1169-L1170

File: test/e2e/middleware-param-rewrite-navigation/pages/a.js (hunk 1 above)
Comment (balazsorban44) [reply test/e2e/middleware-param-rewrite-navigation/pages/a.js:10]: From page "a", you would navigate to "b", so this is correct. (Doesn't really matter though, we target the `a` tag to click.)

Comment (ijjk) [reply test/e2e/middleware-param-rewrite-navigation/pages/a.js:10]: Oh then the other page should be "to a" right? Currently both say "to b" which was a bit confusing when reproducing

**Comment by balazsorban44** [comment]
Closing as not actively working on it, and might be fixed in later `canary` versions.

**Pull request closed without merging**